core/dictgen/src/BaseSelectionRule.cxx

```diff
@@ -229,7 +229,7 @@ BaseSelectionRule::EMatchType BaseSelectionRule::Match(const clang::NamedDecl *d
       if (name_value == name) {
          const_cast<BaseSelectionRule*>(this)->SetMatchFound(true);
          return kName;
-      } else if ( fCXXRecordDecl != (void*)-1 ) {
```

Comment:
There is one more instance of comparison to -1 and one assignment to -1.   Did we double check that after this assignment to -1, we can not get here?

Reply:
Do you mean `fCXXRecordDecl != (void*)-1 || !fCXXRecordDecl`?

Do you have an idea what's wrong with the failing PR tests?

Reply:
> Do you mean fCXXRecordDecl != (void*)-1 || !fCXXRecordDecl?

No.  ```fCXXRecordDecl == (void*)-1 || !fCXXRecordDecl``` is likely safer.

But my question is whether you had extended you review of the code to include looking at:
```
  if (! isTypedefNametoRecordDecl && fCXXRecordDecl !=0 && fCXXRecordDecl != (void*)-1) {
...
            const_cast<BaseSelectionRule*>(this)->fCXXRecordDecl = (clang::CXXRecordDecl*)-1;
```
and whether this case needed to be eliminated or not.

Reply:
> Do you have an idea what's wrong with the failing PR tests?

It is strange.  The old condition meant that it was 'not' taken if the value was -1 and the new conditional mean that the if is now true with -1.   Since that lead to a search that is odd that it leads to not finding things that were found before.   

I guess we had no choice but to really (re)understand what this value means for fCXXRecordDecl

Reply:
We have not reviewed the second occurrence of `(void*)-1`. IIUC, the code is: (a) caching the result of a successful lookup by setting `fCXXRecordDecl` and (b) caching the negative lookup if it was not found to avoid lookup again (via the `-1` value).

@yamaguchi1024 and I are thinking of having a negative cache in the LookupHelper directly. That is, we are thinking of implementing a LRU negative caching policy by allocating a (small) map of strings that were recently looked up and were unsuccessful and we will invalidate this cache every time the value of `cling::Interpreter::getLastTransaction` changes.

This would provide a generic solution to the problem the code with `-1` is solving. We have seen some other code elsewhere which does several failing lookups of the same string.

Reply:
Fair enough.  Thanks.